### PR TITLE
Add an ICompilationUnitSyntax interface to allow for a common way to operate on VB and C# root nodes.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Dustin woudl like this so he can write common helper functions that work on tokens transparently of language.  Right now he has to special case C# and VB's CompilationUnitSyntax nodes so he can properly handle their respective 'EndOfFileToken's. With this approach he could write all of his node/token processing code in a fully language neutral manner.

Given how we structure trivia/tokens, i feel this is an appropriate interface to have.  It really enforces the idea of how the root node of a tree must work, and where any final trivia in the tree will belong.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-compiler @DustinCampbell

**Comment by CyrusNajmabadi** [comment]
Note: the reason for the large change to "public-unshipped" is that all the recent IOperation stuff was added in an unsorted manner.  So when you now run the "add new public symbol" code fixer, it produces a sorted list that looks very different.  But note that it's just 530 removed and 532 added.  The removals are actually just moves, and there are only 2 additions, the new interface and the property in it.

**Comment by DustinCampbell** [comment]
Tagging @mattwar as well (and @gafter specifically)

**Comment by DustinCampbell** [comment]
Also @AnthonyDGreen

**Comment by mattwar** [comment]
:+1:

Maybe a Members collection too?

**Comment by CyrusNajmabadi** [comment]
@mattwar I've added 'Members' as well.  Thanks!

**Comment by CyrusNajmabadi** [comment]
@dotnet-bot retest prtest/mac/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_mac_dbg_unit32/2172/
// Retest reason:

File: src/Compilers/Core/Portable/Syntax/ICompilationUnitSyntax.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+﻿using System;
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Syntax/ICompilationUnitSyntax.cs]: missing copyright header

File: src/Compilers/Core/Portable/Syntax/ICompilationUnitSyntax.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    /// is <code>true</code> will implement this interface.
+    ///
+    /// This interface provides a common way to both easily find the root of a <see cref="SyntaxTree"/>
+    /// given any <see cref="SyntaxNode"/>, as well as a common way for handling the special 
+    /// <see cref="EndOfFileToken"/> that is needed to store all final trivia in a <see cref="SourceText"/>
+    /// that is not owned by any other <see cref="SyntaxToken"/>.
+    /// </summary>
+    public interface ICompilationUnitSyntax
+    {
+        SyntaxList<SyntaxNode> Members { get; }
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Syntax/ICompilationUnitSyntax.cs]: XML doc comment describing which kinds of members should be returned?

**Comment by AnthonyDGreen** [comment]
I've considered suggestions to add interfaces for common aspects of syntax nodes several times throughout the development of the v1 APIs. They led to several open design questions that I never felt had satisfactory answers. I'm curious for your thoughts on.

1. It's very cumbersome to use the interfaces with any existing APIs because all of our APIs are written in terms of SyntaxNode. So if you want to use an ICompilationUnitSyntax with any existing API you have to first cast the object to SyntaxNode and if you want to take the result of any APIs you have to cast to ICompilationUnitSyntax. If we could generally express _intersection types_--a type that is _both_ a SyntaxNode _and_ an ICompilationUnitSyntax--code could efficiently exploit the proposed interfaces without casting. Unfortunately this concept is only exposed (or approximated) in C# when applying multiple base type constraints on a generic type parameter. So if all methods viraly trafficked in T where T : SyntaxNode, T : ICompilationUnitSyntax that would provide the ideal experience. Unfortunately that also means every method using this interface has to be generic and many generic instantiations getting jitted but in principle duplicating the code for each implementing node isn't really much different. In my opinion this is the most important issue with adding interfaces--not fragmenting the API consumption experience.

2. We need some naming conventions to delineate interfaces describing common interfaces within a language (e.g. every C# node that takes a control statement or has an open and close curly) and between languages (e.g. ICommonExpressionSynytax).

3. We need a namespace to put these interfaces in. I don't care for the experience of these interfaces being in the same scope as the concrete types as it adds more clutter to IntelliSense as you're forced to tell completion whether you meant CompilationUnitSyntax or ICompilationUnitSyntax. I also think it's an unpleasant experience for new users of the API in general and as with problem 2 above we might end up with 3 different namespaces for the two different kinds of interfaces. This should be an advanced scenario one "opens up" on demand. 

Perhaps this is a glowing example of a problem better solved by adding duck-typing/dynamic-interfaces to the languages? This is a general problem in .NET - a general solution would benefit everyone and in fact if one existed the proposed code changes would be obsolete, no?

**Comment by DustinCampbell** [comment]
@AnthonyDGreen, this isn't about creating interfaces for every syntax node type, just compilation unit since that one's special. Without this, the IDE must write language-specific code just to find the syntax token to the right of a position because the end of file token is needed to do this correctly. This also could have been solved by having `SyntaxKind.EndOfFileToken` share the same value between languages but it doesn't.

My thoughts:

1. Why is casting an issue for ICompilationUnitSyntax? We already cast to concrete node types all over the place. I don't see how this is different.
2. This is out of the scope of this change. We're thinking of ICompilationUnitSyntax as special like IStructuredTriviaSyntax is. Initially, we even called it IRootSyntaxNode but decided to call it ICompilationUnitSyntax since SyntaxTree has a HasCompilationUnit property.
3. Again, we're not adding namespaces for the whole universe of syntax nodes. I actually think that ship has sailed. IOperation is probably as good as we'll get in this space. We're just adding ICompilationUnitSyntax because it's necessary to get the end of file token for general cross-language tree navigation.

Further discussion on the broader design of interfaces across syntax nodes should probably be started with an issue rather than using this pull request.

**Comment by AnthonyDGreen** [comment]
Ah, that does change things a little. It wasn’t clear from the initial description that the circumstances were very particular to CompilationUnit and that it wasn’t the case that CompilationUnit just happened to be the immediate example. The comparison to IStructuredTriviaSyntax makes more sense to me. I just don’t want this case to be a precedent for adding another interface the next time two nodes are structurally similar between languages and it seems that’s less of a risk now that you’ve better explained the rationale.

It would be nice to see how this API can be used for that scenario –finding the token to the right of the cursor—for historical documentation purposes. It’ll make it more discoverable to other users to understand its purpose/value.

Rather than introduce a new namespace I’d prefer it to live alongside IStructuredTriviaSyntax in the CodeAnalysis namespace. Putting it in a namespace implies that it’s part of a family of types centered around a scenario and its nearest relative is in that namespace. Consider tagging it EditorBrowsable(Hidden/Advanced/Never) or whatever if it’s too niche.

Given the purpose described I’m more hesitant on adding members beyond EndOfFile, like Members, if they aren’t specifically needed for the navigation scenario. Rather, it’s very clear why EoF is special but not clear why Members is also special but ImportsOrUsingDirectives or FileAttributes isn’t or if we should make special considerations for script/REPL CompilationUnits (e.g. LastExpression).

**Comment by CyrusNajmabadi** [comment]
Ok.  I'm going to remove 'members', and just make it clear that this type is purely used so you can work with the EOF token in a language neutral way.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler Can i get another sign off?  Thanks!

**Comment by CyrusNajmabadi** [comment]
> It would be nice to see how this API can be used for that scenario –finding the token to the right of the cursor—for historical documentation purposes. It’ll make it more discoverable to other users to understand its purpose/value.

I'm not sure what you mean.  Do you mean you want a comment in the code showing how to use this?  Or you want an example in this discussion for how the API would be used. 

If the latter then dustin can provide good examples as he was the one having to do deal with this issue when trying to do syntax navigation in a language neutral manner.

**Comment by CyrusNajmabadi** [comment]
> Rather than introduce a new namespace I’d prefer it to live alongside IStructuredTriviaSyntax in the CodeAnalysis namespace. 

I've fixed this up.  It was never intentional to put this in a new namespace.  That only happened due to the way the compiler structures its source code (i.e. it uses folders that don't correspond to namespaces).

**Comment by AnthonyDGreen** [comment]
Cyrus, I just meant add a unit test that does the thing you want. That shows how you get the token next to the position of the cursor and how this interface helps that. It'll be simple but it'll make the change history make sense later.

**Comment by CyrusNajmabadi** [comment]
@AnthonyDGreen I've made changes that show the type of sharing that we can now have thanks to this interface.

**Comment by CyrusNajmabadi** [comment]
@DustinCampbell  will likely be able to make many more changes like this as part of the work he's doing.

File: src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxNode.cs [hunk 3]
```diff
@@ -904,17 +717,17 @@ protected override SyntaxTree SyntaxTreeCore
 
         protected override SyntaxToken FindTokenCore(int position, bool findInsideTrivia)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxNode.cs]: @agocke Can i remove this method?  Is it a breaking change to get rid of an override?  I wasn't sure, so i left it in.

Comment (agocke) [inline src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxNode.cs]: Looks OK to me.

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxNode.cs]: thanks!

**Pull request merged**